synapse/cores/sortedram.py

```diff
@@ -0,0 +1,222 @@
+
+    def _sizeByGe(self, prop, valu, limit=None):
+        return self._sizeByRange(prop, (valu, None), limit=limit)
+
+    def _rowsByGe(self, prop, valu, limit=None):
+        return self._rowsByRange(prop, (valu, None), limit=limit)
+
+    def _sizeByLe(self, prop, valu, limit=None):
+        return self._sizeByRange(prop, (None, valu), limit=limit)
+
+    def _rowsByLe(self, prop, valu, limit=None):
+        return self._rowsByRange(prop, (None, valu), limit=limit)
+
+    def _addRows(self, rows):
+        for row in rows:
+            self.rowsbyid[row[0]].add(row)
+            if isint(row[2]):
+                rowsbyprop = self.irowsbyprop
+            else:
+                rowsbyprop = self.srowsbyprop
+            rowsbyprop[row[1]].add(row)
+            self.rowsbyvalu[ (row[1],row[2]) ].add(row)
+
+    def _delRowsById(self, ident):
+        for row in self.rowsbyid.pop(ident,()):
+            self._delRawRow(row)
+
+    def _delRowsByIdProp(self, iden, prop):
+        rows = [ row for row in self.rowsbyid.get(iden) if row[1] == prop ]
+        [ self._delRawRow(row) for row in rows ]
+
+    def _getRowsByIdProp(self, iden, prop):
+        return [ row for row in self.rowsbyid.get(iden) if row[1] == prop ]
+
+    def _delRowsByProp(self, prop, valu=None, mintime=None, maxtime=None):
+        for row in self.getRowsByProp(prop,valu=valu,mintime=mintime,maxtime=maxtime):
+            self._delRawRow(row)
+
+    def _delRawRow(self, row):
+
+        byid = self.rowsbyid.get(row[0])
+        if byid != None:
+            byid.discard(row)
+
+        if isint(row[2]):
+            rowsbyprop = self.irowsbyprop
+        else:
+            rowsbyprop = self.srowsbyprop
+
+        byprop = rowsbyprop[ row[1] ]
+        byprop.discard(row)
+        if not byprop:
+            rowsbyprop.pop(row[1],None)
+
+        propvalu = (row[1],row[2])
+
+        byvalu = self.rowsbyvalu[propvalu]
+        byvalu.discard(row)
+        if not byvalu:
+            self.rowsbyvalu.pop(propvalu,None)
+
+    def _getRowsById(self, iden):
+        return list(self.rowsbyid.get(iden,()))
+
+    def _getRowsByProp(self, prop, valu=None, mintime=None, maxtime=None, limit=None):
+
+        if valu == None:
+            rows = set(self.irowsbyprop.get(prop, ()))
+            rows.update(self.srowsbyprop.get(prop, ()))
+        else:
+            rows = self.rowsbyvalu.get( (prop,valu) )
+
+        if rows == None:
+            return
+
+        c = 0
+        for row in rows:
+            if mintime != None and row[3] < mintime:
+                continue
+
+            if maxtime != None and row[3] >= maxtime:
+                continue
+
+            yield row
+
+            c +=1
+            if limit != None and c >= limit:
+                break
+
+    def _getSizeByProp(self, prop, valu=None, mintime=None, maxtime=None):
+        if valu == None:
+            rows = set(self.irowsbyprop.get(prop, ()))
+            rows.update(self.srowsbyprop.get(prop, ()))
```

Comment:
Since it's impossible (should be) for a row to be in both, why add the overhead of putting these into a set?  ( which adds a hash tree lookup and dedup for each addition )